### ISSUE TITLE
Two feature requests: cache private python pkg, set pre-run python codes when open a notebook

### ISSUE THREAD

**Issue by qqdaiyu55** [jupyterlite/jupyterlite#461]
**Feature Request 1**
Allow local file path in `piplite_urls`, so I can cache my private python package.

**Feature Request 2**
Run python codes defined in config everytime I opened a new notebook.
This can be really good for cutomization, like pre-install some packages, ignore uncessary warnings, ...

If you thinks these requests are resonable, I can help file a PR.

**Comment by qqdaiyu55** [comment]
ok I saw we already support for local path
```
    def resolve_one_wheel(self, path_or_url):
        """download a single wheel, and copy to the cache"""
        local_path = None
        will_fetch = False
```

**Comment by bollwyvl** [comment]
I've thrown out some thoughts on this in a [number](https://github.com/jupyterlite/jupyterlite/issues/438#issuecomment-1010060484) of [places](https://github.com/jupyterlite/jupyterlite/issues/458#issuecomment-1022532513): I don't think pre-running code is a feature that should go into the jupyterlite core, for the same reasons it's not a feature in other clients, and trying to write code in JSON is pretty not fun... but we're pretty much stuck with JSON as a config language.

For kernels, if an approach were to work, it would ideally match the upstream kernel's behavior, and might not work the same way for different kernels. A quick scan shows:

- for ipykernel
  -  [IPython startup files](https://ipython.readthedocs.io/en/stable/interactive/tutorial.html?#startup-files)...
  - ... will honor the `PYTHONSTARTUP` environment variable  
  - ... but doesn't have any command line options
- for JS (which should probably get to where it reuses/mirrors `ijavascript`)
  - doesn't have any config files...
  - ... or enviornment variables...
  - ... but accepts [`--startup-script`](https://github.com/n-riesco/ijavascript#ijskernel-ijavascript-kernel) path...

But even then I'm uncomfortable adding these mechanisms, as this happens _before_ the user gets access to the system. By keeping things in-script (or -notebook), it raises the chances that they will work correctly in multiple settings, and when they fail, it is easy for the user to see why.

**Comment by qqdaiyu55** [comment]
@bollwyvl  Thanks for reply!
> it would ideally match the upstream kernel's behavior a la #97, and might not work the same way for different kernels. A quick scan shows:

Yes, I should make the option name more specific, like `prerunPythonCodes`.

> for ipykernel, IPython startup files...

I didn't get the idea how it can integrate with jupyterlite. As said in IPython doc, we need to create a startup py file in a specific place, but jupyterlite cannot access file system API.

> By keeping things in-script (or -notebook), it raises the chances that they will work correctly in multiple settings, and when they fail, it is easy for the user to see why.

I don't even think it is a concern. "The guy defining prerun codes know what he is doing". At least we should give user an option that we can do it.

-------------

There are many typical cases that we need to prerun codes

**Case 1**
If I provide a private & common pkg for users, it's painful for them to insert the pkg installation codes for every notebook... Instead, if we can run them in prerun codes, they can just import the package, that would be much more friendly.

**Case 2**
Everytime we import pandas using pyodide, it will come out below warning, so I had to insert two-line import warning ignore codes at the start of the notebook.
![image](https://user-images.githubusercontent.com/9846706/151731391-28808c0c-df60-488b-a8a7-dd3bd34f8b73.png)

**Comment by qqdaiyu55** [comment]
Or maybe, I need to make an extension, which send kernel execution message with prerun codes when every notebook got started.

**Comment by qqdaiyu55** [comment]
Or I can try to implement an extension to do it, though it's not fun (making an extension only for startup codes).

**Comment by jtpio** [comment]
> Or I can try to implement an extension to do it, though it's not fun (making an extension only for startup codes).

I think this was also requested upstream in JupyterLab but can't find the issue right now. So having it as an extension could also be useful to regular JupyterLab users.

> Yes, that would give you your results most quickly. I'd listen for app.serviceManager.sessions.runningChanged, and you'd be able to react to it appropriately.

This would also have the benefit of working with JupyterLab and RetroLab / Notebook v7.

There is also this extension which looks similar but does not address the exact same use case: https://github.com/epi2me-labs/jupyterlab-autorun-cells. Although it can be interesting to look at for inspiration.

**Comment by jtpio** [comment]
> **Case 1**
> e.g. #438
> If I provide a common pkg for users, it's painful for them to insert the pkg installation codes for every notebook... Instead, if we can run them in prerun codes, they can just import the package, that would be much more friendly.

Probably this would be the main advantage for such a feature. Users could avoid running `await piplite.install('mypackage')`.

**Comment by bollwyvl** [comment]
> Users could avoid running

While I agree there are cosmetic benefits under certain conditions (e.g. documentation for a specific package), I think every line of code lite runs that the end user can't easily inspect (either success or fail) decreases the reproducibility of the overall system. I'd like to see us get rid of even more up-front code, such as our patches, and very much didn't want to go messing with `micropip`, but the private registry feature was unavoidable, and unlikely to be merged into the upstream without a lot of back-and-forth. 

As to reproducibility, I mean: 

- as a site user, if I download a notebook from one lite site and try to run it in another site: and they both secretly run different (incompatible) lines of code before I even get to do anything, it's going to be very confusing as to why my code doesn't work.
- as a site owner, I am free to install additional lab- and server-extensions to change the behavior... but the behavior won't then be coming _from_ jupyterlite core

Another direction altogether: if we introduced a general pattern of parameterized "kernel environments," or "profiles" or something which could encapsulate these in an user-inspectable (and -customizable) way and showed up as separate kernels, and reported errors well, then most of my concerns would be allayed. Indeed, we're probably _better_ set up for this than "real" kernels, though there is work underway in e.g. jupyter_server for this, and will account for more of these issues.

**Comment by bollwyvl** [comment]
Right, but as soon as a user on a lite site wants packages that an owner
has not chosen to install for them, they _would_ need to know about
piplite.install.

A lite site _isn't_ a real jupyter experience (or is at least quite
different), and a lot of things won't work the same way, though we are
trying hard to make it closer.

**Comment by martinRenou** [comment]
Supporting the IPython config would be nice indeed! Because we already use IPython it could possibly be done.

On another note, there is a PR related to this issue in replite: https://github.com/jtpio/replite/pull/24

**Comment by michaelweinold** [comment]
If I understand correctly from 

 - https://github.com/jtpio/replite/pull/24
 - https://github.com/jupyterlite/jupyterlite/issues/508

referenced by @martinRenou above, the JupyterLite REPL [allows for custom code to be pre-run](https://jupyterlite.readthedocs.io/en/latest/quickstart/embed-repl.html#auto-execute-code-on-startup), but not "silently".

[As others have noted](https://github.com/jupyterlite/jupyterlite/issues/461#issue-1118400997), it would be very helpful to silently pre-run code outside of the REPL shell. Loading packages into a Pyodide kernel using `micropip` would be one use-case.

I was wondering if there was any chance that the IPython config suggested by @bollwyvl will be supported in the future.

**Comment by jobovy** [comment]
@michaelweinold As mentioned here: https://github.com/jupyterlite/jupyterlite/issues/508#issuecomment-1501345360, silently pre-running code in the REPL can be done with this extension: https://github.com/jobovy/jupyterlite-repl-prerun. This assumes that you put the code to be pre-run in a URL parameter. It wouldn't be difficult to write a similar extension that reads the code to be pre-run from a config file similar to how it would be done using the IPython config.
